pkg/smbios/table.go

```diff
@@ -0,0 +1,161 @@
+}
+
+// GetBytesAt returns a number of bytes from the structured part at the specified offset.
+func (t *Table) GetBytesAt(offset, length int) ([]byte, error) {
+	if offset > len(t.data)-length {
+		return nil, fmt.Errorf("invalid offset %d", offset)
+	}
+	return t.data[offset : offset+length], nil
+}
+
+// GetWordAt returns a 16-bit word from the structured part at the specified offset.
+func (t *Table) GetWordAt(offset int) (res uint16, err error) {
+	if offset > len(t.data)-2 {
+		return 0, fmt.Errorf("invalid offset %d", offset)
+	}
+	err = binary.Read(bytes.NewReader(t.data[offset:offset+2]), binary.LittleEndian, &res)
+	return res, err
+}
+
+// GetDWordAt returns a 32-bit word from the structured part at the specified offset.
+func (t *Table) GetDWordAt(offset int) (res uint32, err error) {
+	if offset > len(t.data)-4 {
+		return 0, fmt.Errorf("invalid offset %d", offset)
+	}
+	err = binary.Read(bytes.NewReader(t.data[offset:offset+4]), binary.LittleEndian, &res)
+	return res, err
+}
+
+// GetQWordAt returns a 64-bit word from the structured part at the specified offset.
+func (t *Table) GetQWordAt(offset int) (res uint64, err error) {
+	if offset > len(t.data)-8 {
+		return 0, fmt.Errorf("invalid offset %d", offset)
+	}
+	err = binary.Read(bytes.NewReader(t.data[offset:offset+8]), binary.LittleEndian, &res)
+	return res, err
+}
+
+// GetStringAt returns a string pointed to by the byte at the specified offset in the structured part.
+// NB: offset is not the string index.
+func (t *Table) GetStringAt(offset int) (string, error) {
+	if offset >= len(t.data) {
+		return "", fmt.Errorf("invalid offset %d", offset)
+	}
+	stringIndex := t.data[offset]
+	switch {
+	case stringIndex == 0:
+		return "", nil
+	case int(stringIndex) <= len(t.strings):
+		return t.strings[stringIndex-1], nil
+	default:
+		return "", fmt.Errorf("invalid string index %d", stringIndex)
+	}
+}
+
+func (t *Table) String() string {
+	lines := []string{
+		t.Header.String(),
+		"\tHeader and Data:",
+	}
+	data := t.data
+	for len(data) > 0 {
+		ld := data
+		if len(ld) > 16 {
+			ld = ld[:16]
+		}
+		ls := make([]string, len(ld))
+		for i, d := range ld {
+			ls[i] = fmt.Sprintf("%02X", d)
+		}
+		lines = append(lines, "\t\t"+strings.Join(ls, " "))
+		data = data[len(ld):]
+	}
+	if len(t.strings) > 0 {
+		lines = append(lines, "\tStrings:")
+		for _, s := range t.strings {
+			lines = append(lines, "\t\t"+s)
+		}
+	}
+	return strings.Join(lines, "\n")
+}
+
+// ParseTable parses a table from byte stream.
+// Returns the parsed table and remaining data.
+func ParseTable(data []byte) (*Table, []byte, error) {
+	var err error
+	var h Header
+	if err = h.Parse(data); err != nil {
+		return nil, data, err
+	}
+	if len(data) < int(h.Length)+2 /* string terminator length */ {
+		return nil, data, errors.New("data too short")
+	}
+	structData := data[:h.Length]
+	data = data[h.Length:]
+	var strings []string
+	for len(data) > 0 && err == nil {
+		end := bytes.IndexByte(data, 0)
```

Comment:
This string reading could become a good uio Lexer function.

Reply:
would it? it's quite specific to smbios, not sure it makes much sense in a more generic context.